DAOS-8620 test: Copy and setenv for D_FI_CONFIG.

rpadma2 opened the pull request:
Signed-off-by: rpadma2 <ravindran.padmanabhan@intel.com>

Summary
- Created the FaultInjection class and updated the test.py to create Faultinjection object.
- The fault injection file is copied to all the test nodes. This is done so that we can enable fault injection at the server level.
- D_FI_CONFIG environment variable is set on the test nodes once the fault file is copied to the nodes.

daosbuild1 commented on src/tests/ftest/util/fault_config_utils.py:
```diff
@@ -211,34 +214,70 @@ class FaultInjectionFailed(Exception):
+        self._hosts = None
+        self._fault_file = fault_file
+
+    def write_fault_file(path, fault_list=None, on_the_fly_fault=None):
```
(pylint-no-self-argument) Method should have "self" as first argument

phender replied:
```suggestion
    def write_fault_file(on_the_fly_fault=None):
```

daosbuild1 commented on src/tests/ftest/util/fault_config_utils.py:
```diff
@@ -211,34 +214,70 @@ class FaultInjectionFailed(Exception):
+
+        return fi_config
+
+    def start(self):
```
(pylint-missing-function-docstring) Missing function or method docstring

phender replied:
```suggestion
    def start(self, fault_list):
```

daosbuild1 commented on src/tests/ftest/util/fault_config_utils.py:
```diff
@@ -211,34 +214,70 @@ class FaultInjectionFailed(Exception):
+            if fault_list:
+                # not using workdir because the huge path was messing up
+                # orterun or something, could re-evaluate this later
+                self._fault_file = self.write_fault_file(self.tmp,
```
(pylint-too-many-function-args) Too many positional arguments for method call
(pylint-no-member) Instance of 'FaultInjection' has no 'tmp' member

daosbuild1 commented on src/tests/ftest/util/fault_config_utils.py:
```diff
@@ -211,34 +214,70 @@ class FaultInjectionFailed(Exception):
+                                                         fault_list,
+                                                         None)
+
+    def copy_fault_files(self, hosts):
```
(pylint-missing-function-docstring) Missing function or method docstring

daosbuild1 commented on src/tests/ftest/util/fault_config_utils.py:
```diff
@@ -211,34 +214,70 @@ class FaultInjectionFailed(Exception):
+    def copy_fault_files(self, hosts):
+        if os.path.exists(self._fault_file):
+            self._hosts = list(hosts)
+            distribute_files(self._hosts, self._fault_file, self.tmp)
```
(pylint-no-member) Instance of 'FaultInjection' has no 'tmp' member

phender replied:
```suggestion
            distribute_files(self._hosts, self._fault_file, self._fault_file)
```

daosbuild1 commented on src/tests/ftest/util/fault_config_utils.py:
```diff
@@ -211,34 +214,70 @@ class FaultInjectionFailed(Exception):
+            self._hosts = list(hosts)
+            distribute_files(self._hosts, self._fault_file, self.tmp)
+
+    def stop(self):
```
(pylint-missing-function-docstring) Missing function or method docstring

daosbuild1 commented on src/tests/ftest/util/fault_config_utils.py:
```diff
@@ -211,34 +214,70 @@ class FaultInjectionFailed(Exception):
+        if fault_list:
+            # Remove the fault injection files on the hosts.
+            rm_command = "{} rm -f {}".format(
+                         get_clush_command(self._hosts, "-S -v", True), self.tmp)
```
(pylint-no-member) Instance of 'FaultInjection' has no 'tmp' member

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-6800/1/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

FYI: Errors found in lines not modified in the patch:

[src/tests/ftest/util/server_utils_params.py:176](https://github.com/daos-stack/daos/blob/886cea678fbe4218e297fe9b01f34fad3a899721/src/tests/ftest/util/server_utils_params.py#L176):
(pylint-consider-using-enumerate) Consider using enumerate instead of iterating with range and len

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-6800/1/execution/node/88/log

phender commented on src/tests/ftest/util/fault_config_utils.py:
```diff
@@ -211,34 +214,70 @@ class FaultInjectionFailed(Exception):
+
+            Returns the name of the file.
+        """
+        if fault_list is None and on_the_fly_fault is None:
```
```suggestion
        if self._fault_list is None and on_the_fly_fault is None:
```

rpadma2 replied:
Done.

phender commented on src/tests/ftest/util/fault_config_utils.py:
```diff
@@ -211,34 +214,70 @@ class FaultInjectionFailed(Exception):
+        if fault_list is None and on_the_fly_fault is None:
+            raise FaultInjectionFailed("bad parameters")
+
+        if not os.path.exists(path):
+            os.makedirs(path)
+        fi_config = path + '/fi.yaml'
```
This requires moving `self.test_dir = os.getenv("DAOS_TEST_LOG_DIR", "/tmp")` from `TestWithoutServers.__init__()` to `Test.__init__()`:
```suggestion
        fi_config = os.path.join(self.test_dir, "fi.yaml")
```

rpadma2 replied:
Done.

phender commented on src/tests/ftest/util/fault_config_utils.py:
```diff
@@ -211,34 +214,70 @@ class FaultInjectionFailed(Exception):
+                      allow_unicode=True)
+
+            fault_config = []
+            if fault_list is not None:
```
```suggestion
            if self._fault_listis not None:
```

phender commented on src/tests/ftest/util/fault_config_utils.py:
```diff
@@ -211,34 +214,70 @@ class FaultInjectionFailed(Exception):
+
+            fault_config = []
+            if fault_list is not None:
+                for fault in fault_list:
```
```suggestion
                for fault in self._fault_list:
```

rpadma2 replied:
Done.

phender commented on src/tests/ftest/util/fault_config_utils.py:
```diff
@@ -211,34 +214,70 @@ class FaultInjectionFailed(Exception):
+            yaml.dump({'fault_config': fault_config}, outfile,
+                      default_flow_style=False, allow_unicode=True)
+
+        return fi_config
```
```suggestion
        self._fault_file = fi_config
```

phender commented on src/tests/ftest/util/fault_config_utils.py:
```diff
@@ -211,34 +214,70 @@ class FaultInjectionFailed(Exception):
+            if fault_list:
+                # not using workdir because the huge path was messing up
+                # orterun or something, could re-evaluate this later
+                self._fault_file = self.write_fault_file(self.tmp,
+                                                         fault_list,
+                                                         None)
```
```suggestion
                self._fault_file = self.write_fault_file(None)
```

rpadma2 replied:
Done.

phender commented on src/tests/ftest/util/fault_config_utils.py:
```diff
@@ -211,34 +214,70 @@ class FaultInjectionFailed(Exception):
+
+    def start(self):
+        # setup fau]lt injection, this MUST be before API setup
+        if self._fault_file is None:
```
```suggestion
```

rpadma2 replied:
Removed.

phender commented on src/tests/ftest/util/fault_config_utils.py:
```diff
@@ -211,34 +214,70 @@ class FaultInjectionFailed(Exception):
+
+    def stop(self):
+        # setup fault injection, this MUST be before API setup
+        fault_list = self.params.get("fault_list", '/run/faults/*')
+        if fault_list:
```
```suggestion
        error_list = []
        if self.fault_file:
```

phender commented on src/tests/ftest/util/fault_config_utils.py:
```diff
@@ -211,34 +214,70 @@ class FaultInjectionFailed(Exception):
+        fault_list = self.params.get("fault_list", '/run/faults/*')
+        if fault_list:
+            # Remove the fault injection files on the hosts.
+            rm_command = "{} rm -f {}".format(
```
```suggestion
            rm_command = "{} rm -f {}".format(get_clush_command(self._hosts, "-S -v", True), self.fault_file)
```

phender commented on src/tests/ftest/util/fault_config_utils.py:
```diff
@@ -211,34 +214,70 @@ class FaultInjectionFailed(Exception):
+            # Remove the fault injection files on the hosts.
+            rm_command = "{} rm -f {}".format(
+                         get_clush_command(self._hosts, "-S -v", True), self.tmp)
+            run_command(rm_command, verbose=True, raise_exception=False)
```
```suggestion
            try:
                run_command(rm_command, verbose=True, raise_exception=False)
            except DaosTestError as error:
                error_list.append(error)
                
        return error_list
```

phender commented on src/tests/ftest/util/fault_config_utils.py:
```diff
@@ -211,34 +214,70 @@ class FaultInjectionFailed(Exception):
-    return fi_config
+    def __init__(self, test, fault_file):
+        super().__init__(test, fault_file)
+        self._test = test
```
```suggestion
```

phender commented on src/tests/ftest/util/fault_config_utils.py:
```diff
@@ -211,34 +214,70 @@ class FaultInjectionFailed(Exception):
+    def start(self):
+        # setup fau]lt injection, this MUST be before API setup
+        if self._fault_file is None:
+            fault_list = self.params.get("fault_list", '/run/faults/*')
```
```suggestion
```

phender commented on src/tests/ftest/util/apricot/apricot/test.py:
```diff
@@ -435,15 +435,11 @@ def setUp(self):
-                                                                  None)
-            os.environ["D_FI_CONFIG"] = self.fault_file
+        self.fault_injection = FaultInjection(self.test_id, None)
+        self.fault_injection.start()
```
```suggestion
        self.fault_injection.start(self.params.get("fault_list", '/run/faults/*'))
```

rpadma2 commented on src/tests/ftest/util/fault_config_utils.py:
```diff
@@ -8,6 +8,9 @@
 import io
 import yaml
+from avocado import Test
+from general_utils import distribute_files, run_command, \
+                          get_clush_command
```
Done.

rpadma2 commented on src/tests/ftest/util/fault_config_utils.py:
```diff
@@ -211,34 +214,70 @@ class FaultInjectionFailed(Exception):
 
-def write_fault_file(path, fault_list=None, on_the_fly_fault=None):
-    """ Write out a fault injection config file.
+class FaultInjection(Test):
```
Done.

rpadma2 commented on src/tests/ftest/util/fault_config_utils.py:
```diff
@@ -8,6 +8,9 @@
 import os
 import io
 import yaml
+from avocado import Test
```
Done.

rpadma2 commented on src/tests/ftest/util/fault_config_utils.py:
```diff
@@ -211,34 +214,70 @@ class FaultInjectionFailed(Exception):
+        return fi_config
+
+    def start(self):
+        # setup fau]lt injection, this MUST be before API setup
```
Done. Typo fixed.

rpadma2 commented on src/tests/ftest/util/apricot/apricot/test.py:
```diff
@@ -452,14 +448,7 @@ def setUp(self):
-                self._teardown_errors.append(
-                    "Error running inherited teardown(): {}".format(error))
-
+        self.fault_injection.stop()
```
Done.

rpadma2 commented on src/tests/ftest/util/fault_config_utils.py:
```diff
@@ -211,34 +214,70 @@ class FaultInjectionFailed(Exception):
-
-    return fi_config
+    def __init__(self, test, fault_file):
+        super().__init__(test, fault_file)
```
Done.

rpadma2 commented on src/tests/ftest/util/fault_config_utils.py:
```diff
@@ -211,34 +214,70 @@ class FaultInjectionFailed(Exception):
+        super().__init__(test, fault_file)
+        self._test = test
+        self._hosts = None
+        self._fault_file = fault_file
```
Done.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[src/tests/ftest/util/server_utils_params.py:176](https://github.com/daos-stack/daos/blob/4815ad8c740345ab2d1361b77545ab5d06dd8cd1/src/tests/ftest/util/server_utils_params.py#L176):
(pylint-consider-using-enumerate) Consider using enumerate instead of iterating with range and len

phender commented on src/tests/ftest/util/fault_config_utils.py:
```diff
@@ -211,34 +218,79 @@ class FaultInjectionFailed(Exception):
+        """
+        # Remove the fault injection files on the hosts.
+        error_list = []
+        rm_command = "{} rm -f {}".format(
+                        get_clush_command(self._hosts, "-S -v", True), self._fault_file)
+        try:
```
For the situation where fault injection is being used with the `TestWithoutServer` class - where the `FaultInjection.copy_fault_files()` method will not be called - this fail because `clush` will be called witha n empty `-w` list.
```suggestion
        commands = []
        if self._host_list:
            commands.append(get_clush_command(self._hosts, "-S -v", True))
        commands.append("rm -f {}".format(self._fault_file))
        try:
            run_command(" ".join(commands), verbose=True, raise_exception=False)
```

rpadma2 replied:
Done.

daosbuild1 commented on src/tests/ftest/util/fault_config_utils.py:
```diff
@@ -211,34 +218,80 @@ class FaultInjectionFailed(Exception):
+        # Remove the fault injection files on the hosts.
+        error_list = []
+        commands = ["rm -f {}".format(self._fault_file)]
+        if self._host_list:
```
(pylint-no-member) Instance of 'FaultInjection' has no '_host_list' member

daosbuild1 reviewed (changes requested):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-6800/3/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

FYI: Errors found in lines not modified in the patch:

[src/tests/ftest/util/server_utils_params.py:176](https://github.com/daos-stack/daos/blob/17cad4d2f7dd4e6254b17801fa04bf59aa955e71/src/tests/ftest/util/server_utils_params.py#L176):
(pylint-consider-using-enumerate) Consider using enumerate instead of iterating with range and len

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-6800/3/execution/node/120/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[src/tests/ftest/util/server_utils_params.py:176](https://github.com/daos-stack/daos/blob/c4964dccc0eee313f2ed2442559c9e3512e2b059/src/tests/ftest/util/server_utils_params.py#L176):
(pylint-consider-using-enumerate) Consider using enumerate instead of iterating with range and len

phender commented on src/tests/ftest/util/server_utils_params.py:
```diff
@@ -453,6 +453,11 @@ def get_params(self, test):
                 self.log.debug("Ignoring the scm_size when scm_class is 'dcpm'")
                 self.scm_size.update(None, "scm_size")
 
+            # Include fault injection settings if configured
+            fault_file = os.environ.get("D_FI_CONFIG")
+            if fault_file and fault_file not in self.env_vars.value:
+                self.env_vars.update("D_FI_CONFIG={}".format(fault_file), "env_vars", True)
```
I was thinking about this approach and I was wondering if there could be a situation where `D_FI_CONFIG` may not be cleared between tests.  I'm not sure if we could get into that situation, but since this method is already provided a `TestWithServers` object (through the `test` argument), we have a more accurate means of determining if fault injection is active: `test.fault_injection._fault_file is not None`.  It will probably makes sense to change `FaultInjection._fault_file` to `FaultInjection.fault_file` to allow 'public' access - at least to avoid the pylint error.
```suggestion
            # Include fault injection settings if configured
            if test.fault_injection.fault_file:
                fault_setting = "D_FI_CONFIG={}".format(test.fault_injection.fault_file)
                if fault_setting not in self.env_vars.value:
                    self.env_vars.update(fault_setting , "env_vars", True)
```

rpadma2 replied:
Sure. Done.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[src/tests/ftest/util/server_utils_params.py:176](https://github.com/daos-stack/daos/blob/9ee3341ab2145fcb163b44c0f61f99b71da5797c/src/tests/ftest/util/server_utils_params.py#L176):
(pylint-consider-using-enumerate) Consider using enumerate instead of iterating with range and len

phender reviewed (commented):
Code changes look good, thanks.  My only concern is I didn't see any tests run in https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-6800/5 that verify the fault injection code changes are working.  A quick search found only these tests using `fault_list:` in their test yaml:

- container/snapshot.yaml (but it's empty: `fault_list: []`)
- dtx/basic.yaml (but it's commented out or empty)
- fault_injection/pool.yaml
- soak/faults.yaml

rpadma2 commented:
> Code changes look good, thanks. My only concern is I didn't see any tests run in https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-6800/5 that verify the fault injection code changes are working. A quick search found only these tests using `fault_list:` in their test yaml:
> 
> * container/snapshot.yaml (but it's empty: `fault_list: []`)
> * dtx/basic.yaml (but it's commented out or empty)
> * fault_injection/pool.yaml
> * soak/faults.yaml

Are these tests run on CI? Do they work under CI? I believe most of the fault injection tests are manual testing. Is it not?

dinghwah commented:
Please include cart_fault_injection/pool_fault_inject.py to the test which is in CI and current failing on the weekly CI.

rpadma2 commented:
> Please include cart_fault_injection/pool_fault_inject.py to the test which is in CI and current failing on the weekly CI.

Are we sure fault injection framework is working? All, we did is convert the present fault injection methods to class. I am not clear how it will affect cart_fault_injection test?

phender commented:
> > Please include cart_fault_injection/pool_fault_inject.py to the test which is in CI and current failing on the weekly CI.
> 
> Are we sure fault injection framework is working? All, we did is convert the present fault injection methods to class. I am not clear how it will affect cart_fault_injection test?

These changes will affect the fault_injection/pool.py (note the recent rename) test as it enables fault injection through the use of the `fault_list:` yaml parameter.  This test did fail in https://build.hpdd.intel.com/job/daos-stack/job/daos/job/weekly-testing/209/artifact/Functional%20Hardware%20Large/cart_fault_injection/pool_fault_inject.py/, but your changes - which also enable the servers to run with the fault injection file - may have an impact.  There's only one way to find out :)

rpadma2 commented:
> > > Please include cart_fault_injection/pool_fault_inject.py to the test which is in CI and current failing on the weekly CI.
> > 
> > 
> > Are we sure fault injection framework is working? All, we did is convert the present fault injection methods to class. I am not clear how it will affect cart_fault_injection test?
> 
> These changes will affect the fault_injection/pool.py (note the recent rename) test as it enables fault injection through the use of the `fault_list:` yaml parameter. This test did fail in https://build.hpdd.intel.com/job/daos-stack/job/daos/job/weekly-testing/209/artifact/Functional%20Hardware%20Large/cart_fault_injection/pool_fault_inject.py/, but your changes - which also enable the servers to run with the fault injection file - may have an impact. There's only one way to find out :)

I don't think we should try to enable some tests which is already broken on this PR. Without knowing what is causing the failure on cart_fault_injection test, enabling it in this PR makes no sense. After taking a look at the log, I notice the exclude command failed and we have no clue why it failed... I would have a separate PR for enabling the cart_fault_injection test after we land this PR.

---Failure seen on the job.log for the car test----
2021-09-25 13:52:17,291 process          L0626 INFO | Running '/usr/bin/dmg -o /etc/daos/daos_control.yml -d pool exclude TestLabel_2 --rank=0'
2021-09-25 13:52:17,429 process          L0458 DEBUG| [stderr] DEBUG 13:52:17.428855 main.go:218: debug output enabled
2021-09-25 13:52:17,429 process          L0458 DEBUG| [stderr] DEBUG 13:52:17.429561 main.go:246: control config loaded from /etc/daos/daos_control.yml
2021-09-25 13:52:17,431 process          L0458 DEBUG| [stderr] DEBUG 13:52:17.431625 pool.go:629: Exclude DAOS pool target request: &{{{} {{0 {0 0 <nil>}  []} 0xf82ca0} {0 0 0 <nil> <nil>}} TestLabel_2 0 []}
2021-09-25 13:52:17,433 process          L0458 DEBUG| [stderr] DEBUG 13:52:17.433862 rpc.go:234: request hosts: [wolf-110:10001 wolf-111:10001 wolf-112:10001 wolf-113:10001 wolf-114:10001]
2021-09-25 13:52:18,514 process          L0458 DEBUG| [stdout] Exclude command failed: pool Exclude failed: DER_NONEXIST(-1005): The specified entity does not exist
2021-09-25 13:52:18,514 process          L0458 DEBUG| [stderr] ERROR: dmg: pool Exclude failed: DER_NONEXIST(-1005): The specified entity does not exist
2021-09-25 13:52:18,517 process          L0714 INFO | Command '/usr/bin/dmg -o /etc/daos/daos_control.yml -d pool exclude TestLabel_2 --rank=0' finished with 1 after 1.222330093383789s
2021-09-25 13:52:18,543 general_utils    L0285 INFO | Error occurred running '/usr/bin/dmg -o /etc/daos/daos_control.yml -d pool exclude TestLabel_2 --rank=0': Command '/usr/bin/dmg -o /etc/daos/daos_control.yml -d pool exclude TestLabel_2 --rank=0' failed.
stdout: b'Exclude command failed: pool Exclude failed: DER_NONEXIST(-1005): The specified entity does not exist\n'

rpadma2 commented:
Unless, we have some tests which passes fault injection, we shoould not add new variables in landing this PR. If some tests are failing we have to handle it separately in a different PR.

rpadma2 commented:
I ran a manual test adding the fault_list to one of the scrubber test which I am writing... It seems to generate the fault file and set the D_FI_CONFIG on all the nodes. Still not sure whether the faults are really enabled by daos infrastructure... Also, feature/csum-scrubbing branch is broken now and I can't test any further... Does soak tests fault injection framework work? If so, it would be nice if we copy the files in this PR and do the soak testing and see whether we haven't break any existing feature.

rpadma2 commented:
[fault_injection_job.log](https://github.com/daos-stack/daos/files/7256117/fault_injection_job.log)

I ran the fault_injection/pool.py test manually. Definitely, the fault injection framework is working in creating necessary faults. I changed some parameters under util/fault_config_utils.py
original value:
    'DAOS_POOL_CREATE_FAIL_CORPC': {
        'id': '65632',
        'probability_x': '100',
        'probability_y': '20',
        'interval': '1',
        'max_faults': '10'}
to
    'DAOS_POOL_CREATE_FAIL_CORPC': {
        'id': '65632',
        'probability_x': '100',
        'probability_y': '100',
        'interval': '1',
        'max_faults': '1'}

The original value of creating faults at 20 percent is not correct. We have to make sure we create faults 100 percent of the time. Also, the code doesn't handle dmg error exceptions... Once a pool create fails, it just terminates... So, I add the following to the fault_injection/pool.py. 
dmg_command.exit_status_exception = True

The above statement is required if the test is expected to continue when dmg command fails. If not, the test will terminate.

Now, the container create/open fails... I believe we have to handle the daos_command similarly to dmg command in handling the exception.  I am pretty sure the test is flawed in how it handle the dmg/daos errors. 

To validate my thesis, I removed the container faults from the pool.yaml file and test continues now with pool create showing a single timeout (which is expected and handled in the script).

Attaching the job.log script on my manual run. I am pretty confident the fault injection framework works and the problem is with the test case which needs rework.

phender reviewed (approved):
Thanks for confirming Ravi.
